### PR TITLE
Block removal of btp-operator resource if SI or SB exists

### PR DESCRIPTION
**Pull request by piotrmiskiewicz**
**Description**

Changes proposed in this pull request:
- existing Service Instance and Bindigns blocks the btp operator deletion

**Related issue(s)**

### CODE REVIEW

File: docs/operations.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
----
+    ---
 title: BTP Manager operations 
 ---
```
Comment (IwonaLanger) [inline docs/operations.md:3]:
```suggestion
---
title: BTP Manager operations 
---
```

File: docs/operations.md [hunk 2]
```diff
@@ -77,11 +77,18 @@ To start the deprovisioning process, use the following command:
 kubectl delete btpoperator {BTPOPERATOR_CR_NAME}
 ```
 
-The command triggers deletion of all Service Bindings, Service Instances and module resources in the cluster.
+The command triggers deletion of module resources in the cluster. By default, existing Service Instances or Service Bindings blocks the deletion. To unblock it you need to remove existing Service Instances and Service Bindings. Then, after reconciliation time (at most 15 min) BTP Operator resource will be gone.
```
Comment (IwonaLanger) [inline docs/operations.md:80]:
```suggestion
The command triggers the deletion of the module resources in the cluster. By default, the existing Service Instances or Service Bindings block the deletion. To unblock it, you must remove the existing Service Instances and Service Bindings. Then, after a maximum of 15-minute reconciliation time, the BTP Operator resource is gone.
```

File: docs/operations.md [hunk 3]
```diff
@@ -77,11 +77,18 @@ To start the deprovisioning process, use the following command:
...
+
+You can force the deletion by adding a label to the BTP Operator resource:
```
Comment (IwonaLanger) [inline docs/operations.md:82]:
```suggestion
You can force the deletion by adding this label to the BTP Operator resource:
```

File: docs/operations.md [hunk 4]
```diff
@@ -77,11 +77,18 @@ To start the deprovisioning process, use the following command:
...
+```
+force-delete: "true"
+```
+If the label exists, all existing Service Instance and Service Bindings will be deleted automatically.
```
Comment (IwonaLanger) [inline docs/operations.md:86]:
```suggestion
If you use the label, all the existing Service Instances and Service Bindings are deleted automatically.
```

File: controllers/btpoperator_controller.go [hunk 5]
```diff
@@ -594,6 +612,38 @@ func (r *BtpOperatorReconciler) handleDeprovisioning(ctx context.Context, cr *v1
...
+
+		if numberOfBindings > 0 || numberOfInstances > 0 {
+			logger.Info(fmt.Sprintf("Existing resources (%d instances and %d bindings) blocks btp operator deletion.", numberOfInstances, numberOfBindings))
+			msg := fmt.Sprintf("All service instances and bindings must be removed: %d instance(s) and %d binding(s)", numberOfInstances, numberOfBindings)
+			logger.Info(msg)
+			if updateStatusErr := r.UpdateBtpOperatorStatus(ctx, cr,
+				types.StateDeleting, ServiceInstancesAndBindingsNotCleaned, msg); updateStatusErr != nil {
+				return updateStatusErr
+			}
+			logger.Info(fmt.Sprintf("%v", cr.Status))
```
Comment (szwedm) [inline controllers/btpoperator_controller.go:633]: seems like a redundant log

File: controllers/btpoperator_controller_test.go [hunk 6]
```diff
@@ -201,6 +202,54 @@ var _ = Describe("BTP Operator controller", Ordered, func() {
 			Eventually(updateCh).Should(Receive(matchState(types.StateReady)))
 			btpServiceOperatorDeployment := &appsv1.Deployment{}
 			Expect(k8sClient.Get(ctx, client.ObjectKey{Name: DeploymentName, Namespace: kymaNamespace}, btpServiceOperatorDeployment)).Should(Succeed())
+		})
+
+		AfterEach(func() {
+			deleteSecret := &corev1.Secret{}
+			Expect(k8sClient.Get(ctx, client.ObjectKey{Namespace: kymaNamespace, Name: SecretName}, deleteSecret)).To(Succeed())
+			//Expect(k8sClient.Get(ctx, client.ObjectKey{Namespace: defaultNamespace, Name: btpOperatorName}, cr)).To(Succeed())
```
Comment (PK85) [inline controllers/btpoperator_controller_test.go:210]: remove comments with code

File: controllers/btpoperator_controller_test.go [hunk 7]
```diff
@@ -201,6 +202,54 @@ var _ = Describe("BTP Operator controller", Ordered, func() {
...
+			Expect(k8sClient.Delete(ctx, deleteSecret)).To(Succeed())
+			k8sClient.Get(ctx, client.ObjectKey{Namespace: defaultNamespace, Name: btpOperatorName}, cr)
+			cr.SetLabels(map[string]string{forceDeleteLabelKey: "true"})
+			k8sClient.Update(ctx, cr)
+			time.Sleep(time.Second)
+			//Expect(k8sClient.Delete(ctx, cr)).To(Succeed())
```
Comment (PK85) [inline controllers/btpoperator_controller_test.go:216]: remove comments with code

File: controllers/btpoperator_controller_test.go [hunk 8]
```diff
@@ -189,7 +190,7 @@ var _ = Describe("BTP Operator controller", Ordered, func() {
 		})
 	})
 
-	Describe("Deprovisioning", func() {
+	Describe("Deprovisioning without forceDelete label", func() {
```
Comment (PK85) [inline controllers/btpoperator_controller_test.go:193]: Maybe just ```Deprovisioning without force-delete label```

File: controllers/btpoperator_controller_test.go [hunk 9]
```diff
@@ -201,6 +202,54 @@ var _ = Describe("BTP Operator controller", Ordered, func() {
...
+
+			Eventually(updateCh).Should(Receive(matchReadyCondition(types.StateDeleting, metav1.ConditionFalse, ServiceInstancesAndBindingsNotCleaned)))
+
+			Expect(k8sClient.Delete(ctx, sbUnstructured)).To(Succeed())
+			Expect(k8sClient.Delete(ctx, siUnstructured)).To(Succeed())
+		})
+
+	})
+
+	Describe("Deprovisioning with forceDelete label", func() {
```
Comment (PK85) [inline controllers/btpoperator_controller_test.go:240]: Maybe just ```Deprovisioning with force-delete label```

File: docs/operations.md [hunk 10]
```diff
@@ -77,11 +77,18 @@ To start the deprovisioning process, use the following command:
...
 
 At first, the deprovisioning process tries to perform the deletion in a hard delete mode. It tries to delete all 
 Service Bindings and Service Instances across all Namespaces. The time limit for the hard delete is 20 minutes. 
 After this time, or in case of an error, the process goes into soft delete mode, which runs deletion of finalizers from Service Instances and Service Bindings.
+
 In order to delete finalizers the reconciler deletes module deployment and webhooks.
 Regardless of mode, in the next step, all SAP BTP Service Operator resources marked with the `app.kubernetes.io/managed-by:btp-manager`
 label are deleted. The deletion process of module resources is based on resources GVKs (GroupVersionKinds) found in [manifests](../module-resources).
```
Comment (PK85) [inline docs/operations.md:94]: Add this new condition reason: ServiceInstancesAndBindingsNotCleaned to https://github.com/kyma-project/btp-manager/blob/c1675d8d329327f27ec9270bf9a999eaec1a6c1b/docs/operations.md#conditions

File: controllers/btpoperator_controller.go [hunk 11]
```diff
@@ -563,6 +577,10 @@ func (r *BtpOperatorReconciler) HandleDeletingState(ctx context.Context, cr *v1a
 		logger.Error(err, "deprovisioning failed")
 		return err
 	}
+	if r.reasonExists(cr, ServiceInstancesAndBindingsNotCleaned) {
```
Comment (szwedm) [inline controllers/btpoperator_controller.go:580]: the reconciler will drop deprovisioning when `force-delete` label is added later to the CR and there isn't the Reason change

**Review by szwedm — APPROVED** [review]
/lgtm

**Review by szwedm — APPROVED** [review]
/lgtm

**Review by IwonaLanger — APPROVED** [review]

**Pull request merged**